packages/app/auth/providers/earthdata-login.ts

```diff
@@ -0,0 +1,115 @@
+    endpoint: string
+}
+/**
+ * A custom OAuth2 Provider for authenticating to Earthdata Login
+ * (https://urs.earthdata.nasa.gov, https://next-auth.js.org/configuration/providers/oauth)
+ */
+export default function EarthdataLoginProvider<P extends EarthdataUser = any>(
+    options: OAuthUserConfig<P>
+): OAuthConfig<P> {
+    return {
+        id: 'earthdata-login',
+        name: 'Earthdata Login',
+        version: '2.0',
+        type: 'oauth',
+        authorization: {
+            // https://urs.earthdata.nasa.gov/documentation/for_integrators/api_documentation#GET/oauth/authorize
+            url: `${basePath}/oauth/authorize`,
+            params: {
+                // disables the URS "splash screen" (the one that says "Redirecting...") for logged in users
+                splash: false,
+            },
+        },
+        token: {
+            /**
+             * Earthdata's OAuth2 token retrieval deviates from the Next Auth specification
+             * so we will send out our own POST token request
+             */
+            async request({ params, provider }) {
+                // https://urs.earthdata.nasa.gov/documentation/for_integrators/api_documentation#/oauth/token
+                const result = await fetch(`${basePath}/oauth/token`, {
+                    method: 'POST',
+                    body: `grant_type=authorization_code&code=${
+                        params.code
+                    }&redirect_uri=${encodeURIComponent(provider.callbackUrl)}`,
+                    headers: {
+                        'Content-Type': 'application/x-www-form-urlencoded',
+                        Authorization: Buffer.from(
+                            `${process.env.AUTH_CLIENT_ID}:${process.env.AUTH_CLIENT_SECRET}`
+                        ).toString('base64'),
+                    },
+                })
+
+                const {
+                    access_token,
+                    endpoint,
+                    expires_in,
+                    refresh_token,
+                    token_type,
+                }: EDLTokenSetParameters = await result.json()
+
+                return {
+                    tokens: {
+                        access_token,
+                        endpoint,
+                        expires_at: expires_in,
+                        refresh_token,
+                        token_type,
+                    },
+                }
+            },
+        },
+        userinfo: {
+            async request(context) {
+                // https://urs.earthdata.nasa.gov/documentation/for_integrators/api_documentation#/api/users/%7Buserid%7D
+                const user = await fetch(`${basePath}${context.tokens.endpoint}`, {
+                    headers: {
+                        Authorization: `Bearer ${context.tokens.access_token}`,
+                    },
+                })
+
+                return await user.json()
+            },
+        },
+        profile(profile: EarthdataUser) {
```

Comment:
EDL uses different fields for profile information (ex: first_name instead of NextAuth's firstName). This just maps EDL to NextAUth